hoomd/md/pytest/test_pair.py

```diff
@@ -0,0 +1,751 @@
+    snap = sim.state.snapshot
+    if snap.exists:
+        pos = snap.particles.position
+        mag_r1 = np.linalg.norm(pos[0] - pos[1])
+
+        Fa = -1 * ((E1[0] - E0[0]) / (mag_r1 - mag_r0)) * 2 * direction
+        Fb = -1 * ((E1[1] - E0[1]) / (mag_r1 - mag_r0)) * 2 * direction * -1
+    snap = sim.state.snapshot
+    if snap.exists:
+        snap.particles.position[1] = initial_pos[1]
+    sim.state.snapshot = snap
+    if sim.state.snapshot.exists:
+        return Fa, Fb
+    else:
+        return 0, 0  # return dummy values if not on rank 1
+
+
+def test_force_energy_relationship(simulation_factory,
+                                   two_particle_snapshot_factory,
+                                   valid_params):
+    # don't really test DPD and DPDLJ for this test
+    pot_name = valid_params.pair_potential.__name__
+    if pot_name == "DPD" or pot_name == "DPDLJ":
+        pytest.skip("Cannot test force energy relationship for " +
+                    pot_name + " pair force")
+
+    pair_keys = valid_params.pair_potential_params.keys()
+    particle_types = list(set(itertools.chain.from_iterable(pair_keys)))
+    pot = valid_params.pair_potential(**valid_params.extra_args,
+                                      nlist=hoomd.md.nlist.Cell(),
+                                      r_cut=2.5, mode='none')
+    for pair in valid_params.pair_potential_params:
+        pot.params[pair] = valid_params.pair_potential_params[pair]
+
+    snap = two_particle_snapshot_factory(particle_types=particle_types, d=1.5)
+    _update_snap(valid_params.pair_potential, snap)
+    sim = simulation_factory(snap)
+    integrator = hoomd.md.Integrator(dt=0.005)
+    integrator.forces.append(pot)
+    integrator.methods.append(hoomd.md.methods.Langevin(hoomd.filter.All(),
+                                                        kT=1, seed=1))
+    sim.operations.integrator = integrator
+    sim.operations.schedule()
+    for pair in valid_params.pair_potential_params:
+        snap = sim.state.snapshot
+        if snap.exists:
+            snap.particles.typeid[0] = particle_types.index(pair[0])
+            snap.particles.typeid[1] = particle_types.index(pair[1])
+        sim.state.snapshot = snap
+
+        calculated_forces = _calculate_force(sim)
+        sim_forces = sim.operations.integrator.forces[0].forces
+        if sim_forces is not None:
+            np.testing.assert_allclose(calculated_forces[0],
+                                       sim_forces[0],
+                                       rtol=1e-06)
+            np.testing.assert_allclose(calculated_forces[1],
+                                       sim_forces[1],
+                                       rtol=1e-06)
+
+FandEtuple = namedtuple('FandEtuple',
+                        ['pair_potential',
+                         'pair_potential_params',
+                         'extra_args',
+                         'forces',
+                         'energies'])
+
```

Comment:
Since we are using a dictionary now, we could just keep that representation, but up to you. I don't mind either way.